src/subtyping.md

```diff
@@ -1,146 +1,143 @@
 # Subtyping and Variance
 
-Subtyping is a relationship between types that allows statically typed
-languages to be a bit more flexible and permissive.
+Rust uses lifetimes to track the relationships between borrows and ownership.
+However, a naive implementation of lifetimes would be either too restrictive,
+or permit undefined behaviour.
 
-Subtyping in Rust is a bit different from subtyping in other languages. This
-makes it harder to give simple examples, which is a problem since subtyping,
-and especially variance, is already hard to understand properly. As in,
-even compiler writers mess it up all the time.
+In order to allow flexible usage of lifetimes
+while also preventing their misuse, Rust uses a combination of **Subtyping** and **Variance**.
 
-To keep things simple, this section will consider a small extension to the
-Rust language that adds a new and simpler subtyping relationship. After
-establishing concepts and issues under this simpler system,
-we will then relate it back to how subtyping actually occurs in Rust.
+Let's start with a example.
 
-So here's our simple extension, *Objective Rust*, featuring three new types:
-
-```rust
-trait Animal {
-    fn snuggle(&self);
-    fn eat(&mut self);
+```rust,ignore
+fn debug<T: std::fmt::Debug>(a: T, b: T) {
+    println!("a = {:?} b = {:?}", a, b);
 }
 
-trait Cat: Animal {
-    fn meow(&self);
+fn main() {
+    let hello: &'static str = "hello";
+    {
+        let world = String::from("world");
+        let world = &world; // 'b has a shorter lifetime than 'static
+        debug(hello, world);
+    }
 }
+```
 
-trait Dog: Animal {
-    fn bark(&self);
-}
+In a conservative implementation of lifetimes, since `a` and `b` have differeing lifetimes,
+we might see the following error:
+
+```text
+error[E0308]: mismatched types
+ --> src/main.rs:10:16
+   |
+10 |         debug(hello, world);
+   |                      ^
+   |                      |
+   |                      expected `&'static str`, found struct `&'b str`
 ```
 
-But unlike normal traits, we can use them as concrete and sized types, just like structs.
+This would be rather unfortunate. In this case,
+what we want is to accept any type that lives *at least as long* as `'b`.
+Let's try using subtyping with our lifetimes.
```

Comment:
Hmm. This is contradictory to how the original page represents `&'a T` as a type constructor with type arguments `'a` and `T`

Reply:
> Nit: subtyping applies to types, not lifetimes

I'm not sure that's really a nitpick. I think it's potentially the crux of why this chapter explaining lifetime variance has been quite unintuitive (at least for me) - since it tries too hard to compare lifetime variance to subtyping which is a different thing.

I've tried to expand on this here: https://github.com/rust-lang/nomicon/issues/339#issuecomment-1422694052 as well as link back to other issues that discussed how rust should define variance for lifetimes.